crimsobot/utils/image.py

```diff
@@ -473,4 +372,229 @@ async def acid(ctx: Context, window: int, user_input: Optional[str]) -> BytesIO:
+
+    return img
+
+
+@c.executor_function
+def process_lower_level(img: Image.Image, effect: str, arg: int) -> BytesIO:
+    # this will only loop once for still images
+    frame_list, durations = [], []
+
+    # if a GIF loops, it will have the attribute loop = 0; if not, then attribute does not exist
+    try:
+        img.info['loop']
+        image_loop = True
+    except KeyError:
+        image_loop = False
+        pass
+
+    for _ in ImageSequence.Iterator(img):
+        # if not animated, will throw KeyError
+        try:
+            duration = img.info['duration']  # type: int
+            durations.append(duration)
+        except KeyError:
+            # an empty tuple for durations tells image_to_buffer that image is still
+            pass
+
+        function_dict: Mapping[str, Callable] = {
+            'acid': make_acid_img,
+            'aenima': make_aenima_img,
+            'lateralus': make_lateralus_img,
+            'needban': make_needban_img,
+            'needping': make_needping_img,
+            'pingbadge': make_pingbadge_img,
+            'xokked': make_xokked_img,
+            'resize': resize_img,
+        }
+
+        # these are no longer coroutines
+        img_out = function_dict[effect](img.convert('RGBA'), arg)  # type: ignore
+        frame_list.append(img_out)
+
+    fp = image_to_buffer(frame_list, tuple(durations), image_loop)
+
+    return fp
+
+
+async def process_image(ctx: Context, image: Optional[str], effect: str, arg: Optional[int] = None) -> Any:
+    # grab user image and covert to RGBA
+    img = await fetch_image(ctx, image)
+
+    # if GIF, kick out if too many frames; if not, charge per frame
+    frame_limit = 200
+    cost_per_frame = 0.06
+
+    if (getattr(img, 'is_animated', False)):
+        if img.n_frames > frame_limit:
+            embed = c.crimbed(
+                title=None,
+                descr='Too many frames! Limit is {}'.format(frame_limit),
+                color_name='orange',
+            )
+            await ctx.send(embed=embed, delete_after=15)
+            return None
+        else:
+            cost = img.n_frames * cost_per_frame
+            bal = await crimsogames.check_balance(ctx.author)
+            if bal < cost:
+                embed = c.crimbed(
+                    title="**GIFs ain't free!**",
+                    descr='\n'.join([
+                        "You can't afford this! GIFs cost **\u20A2{:.2f}** per frame.".format(cost_per_frame),
+                        'GIF cost: **\u20A2{:.2f}** · Your balance: **\u20A2{:.2f}**'.format(cost, bal),
+                    ]),
+                    footer='Play >daily, >cringo, >guessmoji, or >emojistory to win crimsoCOIN!',
+                    thumb_name='weary',
+                    color_name='orange',
+                )
+                await ctx.send(embed=embed, delete_after=20)
+                return None
+            else:
+                # debit the user, credit the bot
```

Comment:
love this feature but it's a bit out of scope for a function named `process_image` - i'll let it go for now, imo this whole utility file could be refactored into a better class/OOP thing and we can do it right
do not block on this comment. just wanted to call it out.